### PR TITLE
AM-4393 enroll validation

### PR DESCRIPTION
**Pull request by lgw-gravitee**
fixes AM-4393

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java [hunk 1]
```diff
@@ -277,8 +278,7 @@ private void saveEnrollment(RoutingContext routingContext) {
             // redirect to the original request
             redirectToAuthorize(routingContext);
         } else {
-            // parameters are invalid
-            routingContext.fail(400);
+            routingContext.fail(new EnrollmentChannelValidationException("Validation failed"));
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java:281]: **nickpick:** not sure about the message here :thinking: but I do not have any better idea than "Invalid Parameters" :man_shrugging:

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+public class MFAEnrollFailureHandler extends AbstractErrorHandler {
+    private static final String ERROR_CODE_VALUE = "enrollment_channel_invalid";
+    public MFAEnrollFailureHandler() {
+        super(RootProvider.PATH_MFA_ENROLL);
+    }
+
+    @Override
+    public void doHandle(RoutingContext routingContext) {
+        handleException(routingContext, routingContext.failure());
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java:27]: **question**: is there anything we have to remove from the session to have a fresh enrollment state ?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java [hunk 3]
```diff
@@ -468,7 +464,8 @@ protected void doStart() throws Exception {
                 .handler(clientRequestParseHandler)
                 .handler(redirectUriValidationHandler)
                 .handler(localeHandler)
-                .handler(new MFAEnrollEndpoint(factorManager, thymeleafTemplateEngine, userService, domain, applicationContext));
```
[the remark this answers is not in the document]
Comment (lgw-gravitee) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:471]: yeap, but this is ErrorHandler instance with routing to /error page. Here, we'd like to redirect back to /enroll

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    private void handleException(RoutingContext context, Throwable throwable) {
+        String errorDescription = throwable == null ? "MFA Enrollment failed for unexpected reason" : throwable.getMessage();
+        final MultiMap queryParams = updateQueryParams(context, errorDescription);
+        String errorPageURL = UriBuilderRequest.resolveProxyRequest(context.request(), context.get(CONTEXT_PATH) + errorPage, queryParams, true);
```
[the remark this answers is not in the document]
Comment (lgw-gravitee) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java:48]: thanks Rafal. I fixed that.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java (hunk 2 above)
Comment (lgw-gravitee) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java:43]: I dont think so. Any session variables are set if validation passes

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties [hunk 5]
```diff
@@ -138,6 +138,7 @@ mfa_enroll.fido.description=S
 mfa_enroll.fido=S�lectionnez le bouton "Suivant" pour d�clencher le processus d'enregistrement
 mfa_enroll.on.select.title=Authentification multi-facteurs
 mfa_enroll.on.select.description=Configurer l'authentification multifacteur en choisissant la m�thode pour valider votre identit�
+mfa_enroll.error.default.message=Enrollment failed
```
Comment (tcompiegne) [inline gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties:141]: Erreur d'enregistrement de l'authentification multifacteur

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties (hunk 5 above)
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties:141]:
```suggestion
mfa_enroll.error.default.message=Erreur d'enregistrement de l'authentification multifacteur
```

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:471]: In that case, I guess you can remove the one line 893.

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties (hunk 5 above)
Comment (lgw-gravitee) [reply gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties:141]: thanks

**Review by leleueri — APPROVED** [review]

**Review by leleueri — APPROVED** [review]

**Pull request merged**